### PR TITLE
Implement joint session history

### LINKED ISSUE #11669: Implement joint session history
https://html.spec.whatwg.org/multipage/browsers.html#joint-session-history

The current idea here is to create a `HistoryIterator` that will lazily interleave `Frame` history depending on the `NavigartionDirection`. Each `Frame` entry will now also hold an `Instant` so the order of navigations can be determined.

cc @asajeffrey

### PR DESCRIPTION
**Pull request by cbrewster**
r? @asajeffrey 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11669 (github issue number if applicable).
- [X] These changes do not require tests because I am not sure how to write a test for this (this will be possible once the History API is in place).

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11670)

### CODE REVIEW

**Comment by asajeffrey** [comment]
IRC chat: http://logs.glob.uno/?c=mozilla%23servo&s=8+Jun+2016&e=8+Jun+2016#c451272

**Comment by asajeffrey** [comment]
TL;DR: to implement go(+delta) use an iterator over the frame tree to find the timestamp for the specified entry, then ffwd every frame to that timestamp (which is a no-op for frames already at that timestamp).

**Comment by asajeffrey** [comment]
Also, there may be a spec issue about https://html.spec.whatwg.org/multipage/browsers.html#traverse-the-history-by-a-delta, which seems to imply that it's only one browsing context that changes its current entry.

**Comment by asajeffrey** [comment]
IRC chat: http://logs.glob.uno/?c=mozilla%23servo&s=9+Jun+2016&e=9+Jun+2016#c452402

TL;DR: the spec at https://html.spec.whatwg.org/multipage/browsers.html#joint-session-history says that only fully active documents contribute to the joint session history, which doesn't appear to be in line with existing implementations.

**Comment by asajeffrey** [comment]
Currently still trying to work out what spec to implement!

**Comment by asajeffrey** [comment]
See https://github.com/ConnorGBrewster/ServoNavigation/

**Comment by ConnorGBrewster** [comment]
@asajeffrey I have been thinking through some ideas. It seems to me that if we could get the `HistoryIterator` to iterate across the entire joint session history(prev and next combined), we could get something pretty close to the spec. My idea is to create a `FrameIterator` and it basically is just an iterator that combines a Frame's prev, current, and next into a single iterator. You then store those in the `HistoryIterator` and merge them together chronologically. The only issue I can't get past is step 2 on https://html.spec.whatwg.org/multipage/browsers.html#traverse-the-history-by-a-delta. We have to be able to get the index of the `current entry of the joint session history`. But that is pretty hard since the whole joint session history is lazily merged. Maybe this sparks some ideas for you? I think this directional moves the implementation a lot closer to the spec

**Comment by ConnorGBrewster** [comment]
@asajeffrey hmm at this point I cannot figure out a non-hackish way to get this ordering to work. Would there be any issues with an in-memory representation. From our testing, it seems like there is no way for a `Frame` that already has history to be added to the frame tree, and if a `Frame` is removed from the tree, you can just remove any entries that reference that `Frame` from the in-memory representation. Any objections?

**Comment by asajeffrey** [comment]
The main issue with the in-memory representation is how to deal with a session history that only contains fully active documents. If we're going for all documents contributing to the joint session history then in-memory should be fine.

**Comment by ConnorGBrewster** [comment]
I am closing this until we come up with a solid implementation plan.
